### PR TITLE
Enable ENABLE_ENHANCED_RESOURCE_SCOPING by default

### LINKED ISSUE #49696: Enable ENABLE_ENHANCED_RESOURCE_SCOPING by default in 1.22
Should we promote this now?

If so, we will need to change the default to `true`, set it in older compatibility versions, and write a `precheck` to warn any users that have `discoverySelectors` set in their mesh config

cc @hzxuzhonghu @kfaseela

### PR DESCRIPTION
**Pull request by leosarra**
**Please provide a description of this PR:**

Fixes [#49696](https://github.com/istio/istio/issues/49696)

### CODE REVIEW

**Comment by leosarra** [comment]
I still need to handle the precheck on the configmap

**Comment by leosarra** [comment]
Sorry, I removed the draft status by mistake.
Please wait before reviewing this

File: istioctl/pkg/precheck/precheck.go [hunk 1]
```diff
@@ -179,6 +187,66 @@ func checkExternalNameAlias(cli kube.CLIClient, messages *diag.Messages) error {
...
+			}
+		}
+		scopingImpacted = meshData["discoverySelectors"] != nil
+
+		// Check if mitigation is already in place
+		for _, container := range deployment.Spec.Template.Spec.Containers {
+			if container.Name == "discovery" {
+				for _, envVar := range container.Env {
+					if envVar.Name == "ENHANCED_RESOURCE_SCOPING" && envVar.Value == "true" {
+						scopingImpacted = false
```
Comment (hzxuzhonghu) [inline istioctl/pkg/precheck/precheck.go:223]: can be optimized. We donot need to check env when meshData["discoverySelectors"] == nil

Comment (leosarra) [reply istioctl/pkg/precheck/precheck.go:224]: Added early skip logic

**Review by costinm — APPROVED** [review]
To clarify: this is a double-opt-in feature, meaning that even if it is on by default via env variable, user still needs to update meshConfig and use the discoverySelectors.

Use of discoverySelectors is not something end users will interact with - but it's mainly for vendor supporting multi-tenant Istio. 

If this is true - I think it is a very reasonable change and progression for this feature.

**Comment by leosarra** [comment]
> To clarify: this is a double-opt-in feature, meaning that even if it is on by default via env variable, user still needs to update meshConfig and use the discoverySelectors.
> 
> Use of discoverySelectors is not something end users will interact with - but it's mainly for vendor supporting multi-tenant Istio.
> 
> If this is true - I think it is a very reasonable change and progression for this feature.

Yes you are correct.

Enabling this flag by default will not impact users that do not use discoverySelectors as they are not set by default.
However, users/vendors employing discoverySelectors for soft-multitenancy will notice a change in default behavior.
DiscoverySelectors will now filter custom resources as well (and influence root-ca-creation).

This shift in default behavior is desirable for discoverySelectors users as it ensures that tenants remain unaffected by the custom resources of other tenant namespaces of which he is not supposed to be aware of.

**Comment by costinm** [comment]
I am less concerned about vendors using discoverySelectors - or even
advanced users using soft multitenancy.
It's part of their job - and risks they take by using Istio in a
non-standard way. OSS has minimal e2e or other testing
for the special deployment modes - they are not user-facing features.

It is unfortunate we are mixing user-facing features in MeshConfig with
vendor-facing ones, but that's what we have.

File: releasenotes/notes/enable-enhanced-scope.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+apiVersion: release-notes/v2
+kind: feature
+area: security
+releaseNotes:
+- |
+  **Updated** default value of the feature flag `ENABLE_AUTO_ENHANCED_RESOURCE_SCOPING` to true.
+upgradeNotes:
+- title: Default value of the feature flag `ENHANCED_RESOURCE_SCOPING` to true
+  content: |
+    `ENHANCED_RESOURCE_SCOPING` is enabled by default. This means that meshConfig.discoverySelectors
```
Comment (linsun) [inline releasenotes/notes/enable-enhanced-scope.yaml:10]: Can you clarify a few points?

1. CustomResource meaning Istio resources?
2. limit configurations ....meaning "only process based on the scope of what is specified from meshConfig.discoverySelectors?

**Comment by linsun** [comment]
Agreed this is useful to enable by default. Left some comments to request the release note to be more clear.

File: releasenotes/notes/enable-enhanced-scope.yaml (hunk 2 above)
Comment (leosarra) [reply releasenotes/notes/enable-enhanced-scope.yaml]: I update the release notes to make it more clear

**Comment by leosarra** [comment]
Full-scope testing has passed and I have improved the release notes.
I have also reverted my fix for #49786 as John is working on a fix that does not rely on reflection (see https://github.com/istio/istio/pull/49794). I will rebase the commit once the fix is merged and remove the draft status to this PR.

**Comment by howardjohn** [comment]
> Full-scope testing has passed and I have improved the release notes. I have also reverted my fix for #49786 as John is working on a more robust that does not rely on reflection (see #49794). I will rebase the commit once the fix is merged and remove the draft status to this PR.

FWIW my PR is not just about not using reflection, the fix here makes the filter skipped

**Pull request merged**